pkgs/build-support/rust/fetch-cargo-tarball/default.nix

```diff
@@ -1,116 +1,150 @@
-    fi
-
-    ${cargoUpdateHook}
-
-    # Override the `http.cainfo` option usually specified in `.cargo/config`.
-    export CARGO_HTTP_CAINFO=${cacert}/etc/ssl/certs/ca-bundle.crt
-
-    if grep '^source = "git' Cargo.lock; then
-        echo
-        echo "ERROR: The Cargo.lock contains git dependencies"
-        echo
-        echo "This is currently not supported in the fixed-output derivation fetcher."
-        echo "Use cargoLock.lockFile / importCargoLock instead."
-        echo
-
-        exit 1
-    fi
-
-    cargo vendor $name --respect-source-config | cargo-vendor-normalise > $CARGO_CONFIG
-
-    # Create an empty vendor directory when there is no dependency to vendor
-    mkdir -p $name
-    # Add the Cargo.lock to allow hash invalidation
-    cp Cargo.lock.orig $name/Cargo.lock
-
-    # Packages with git dependencies generate non-default cargo configs, so
-    # always install it rather than trying to write a standard default template.
-    install -D $CARGO_CONFIG $name/.cargo/config;
-
-    runHook postBuild
-  '';
-
-  # Build a reproducible tar, per instructions at https://reproducible-builds.org/docs/archives/
-  installPhase = ''
-    tar --owner=0 --group=0 --numeric-owner --format=gnu \
-        --sort=name --mtime="@$SOURCE_DATE_EPOCH" \
-        -czf $out $name
-  '';
-
-  inherit (hash_) outputHashAlgo outputHash;
-
-  impureEnvVars = lib.fetchers.proxyImpureEnvVars ++ [ "NIX_CRATES_INDEX" ];
-} // (builtins.removeAttrs args [
-  "name" "sha256" "cargoUpdateHook" "nativeBuildInputs"
-]))
+{
+  pname ? null,
+  version ? null,
+  name ? if args ? pname && args ? version then "${pname}-${version}" else "cargo-deps",
+  src ? null,
+  srcs ? [ ],
+  patches ? [ ],
+  sourceRoot ? "",
+  cargoUpdateHook ? "",
+  nativeBuildInputs ? [ ],
+  ...
+}@args:
+# `name` should be mutually exclusive with `pname`
+assert args ? pname -> !(args ? name);
+# `pname` should only be provided with `version` (and vice versa)
+assert args ? pname == args ? version;
+let
+  hash_ =
+    if args ? hash then
+      {
+        outputHashAlgo = if args.hash == "" then "sha256" else null;
+        outputHash = args.hash;
+      }
+    else if args ? sha256 then
+      {
+        outputHashAlgo = "sha256";
+        outputHash = args.sha256;
+      }
+    else
+      throw "fetchCargoTarball requires a hash for ${name}";
+in
+stdenv.mkDerivation (
+  {
+    name = "${name}-vendor.tar.gz";
+    nativeBuildInputs = [
+      cacert
+      git
+      cargo-vendor-normalise
+      cargo
+    ] ++ nativeBuildInputs;
+
+    buildPhase = ''
+      runHook preBuild
+
+      # Ensure deterministic Cargo vendor builds
+      export SOURCE_DATE_EPOCH=1
+
+      if [[ ! -f Cargo.lock ]]; then
+          echo
+          echo "ERROR: The Cargo.lock file doesn't exist"
+          echo
+          echo "Cargo.lock is needed to make sure that cargoHash/cargoSha256 doesn't change"
+          echo "when the registry is updated."
+          echo
+
+          exit 1
```

Comment:
In Nix files, I do like my shell to have two-space indent. `.editorconfig` says four, but @emilazy and I might connive to effect a two-space-coup on shell.

Reply:
If you hide whitespace (https://github.com/NixOS/nixpkgs/pull/332975/files?diff=unified&w=1) you'll see I didn't substantively touch these lines at all. These lines are indented in the diff, but only by the same amount that the surroundings are indented (and was done by `nixfmt`, not manually).

Reply:
Oh yes, this is just me noticing formatting due to, well, the formatting being impacted with `nixfmt`.